### PR TITLE
Better printing of Pair and Dict using typeinfo

### PR DESCRIPTION
**Pull request by rfourquet**
This is a couple small incremental improvements with the typeinfo property, e.g.:
```julia
julia> show(Pair{Integer,Integer}[1=>2]) # BEFORE
Pair{Integer,Integer}[Pair{Integer,Integer}(1, 2)]

julia> show(Dict{Integer,Integer}(1=>2))
Dict{Integer,Integer}(Pair{Integer,Integer}(1, 2))

julia> show(Pair{Integer,Integer}[1=>2]) # AFTER
Pair{Integer,Integer}[1=>2]

julia> show(Dict{Integer,Integer}(1=>2))
Dict{Integer,Integer}(1=>2)
```

### CODE REVIEW

File: base/arrayshow.jl [hunk 1]
```diff
@@ -480,12 +480,21 @@ function typeinfo_prefix(io::IO, X)
     # what the context already knows about the eltype of X:
     eltype_ctx = typeinfo_eltype(typeinfo)
     eltype_X = eltype(X)
-    # Types hard-coded here are those which are created by default for a given syntax
-    if eltype_X == eltype_ctx || !isempty(X) && eltype_X in (Float64, Int, Char, String)
-        ""
-    elseif print_without_params(eltype_X)
-        string(unwrap_unionall(eltype_X).name) # Print "Array" rather than "Array{T,N}"
+    if X isa AbstractDict
+        K, V = typeof(X).parameters
```
Comment (nalimilan) [inline base/arrayshow.jl]: This is only guaranteed to work for `Dict` AFAIK. Use `keytype` and  `valuetype` instead.

File: base/arrayshow.jl [hunk 2]
```diff
@@ -480,12 +480,21 @@ function typeinfo_prefix(io::IO, X)
...
+        if eltype_X == eltype_ctx || !isempty(X) && isconcretetype(K) && isconcretetype(V)
+            string(typeof(X).name)
```
Comment (nalimilan) [inline base/arrayshow.jl]: Are you sure `unwrap_unionall` isn't needed as in the other branch? I don't remember why it's there.

File: base/arrayshow.jl (hunk 1 above)
Comment (rfourquet) [reply base/arrayshow.jl]: Thanks!

File: base/arrayshow.jl (hunk 2 above)
Comment (rfourquet) [reply base/arrayshow.jl]: No I'm not sure... I still need to see an example where this make a difference, to better understand, so I preferred for now to not change the logic from what it was in the Dict code.

File: test/show.jl [hunk 3]
```diff
@@ -730,8 +730,8 @@ let x = [], y = [], z = Base.ImmutableDict(x => y)
     push!(x, y)
     push!(y, x)
     push!(y, z)
-    @test replstr(x) == "1-element Array{Any,1}:\n Any[Any[Any[#= circular reference @-2 =#]], Base.ImmutableDict(Any[Any[#= circular reference @-3 =#]]=>Any[#= circular reference @-2 =#])]"
-    @test repr(z) == "Base.ImmutableDict(Any[Any[Any[#= circular reference @-2 =#], Base.ImmutableDict(#= circular reference @-3 =#)]]=>Any[Any[Any[#= circular reference @-2 =#]], Base.ImmutableDict(#= circular reference @-2 =#)])"
```
Comment (rfourquet) [inline test/show.jl]: This change is not really wanted, but is necessary to make the tests pass. I think it's due to the fact that `Any` is generally the default value for `typeinfo`, but we may instead nead to use `nothing` (or `missing`?). I will investigate and open a separate PR if needed.

**Review by nalimilan — APPROVED** [review]
Looks OK but I'm not really an expert of that code.

File: test/show.jl (hunk 3 above)
Comment (rfourquet) [reply test/show.jl]: So this is due to the fact that we show `Dict` easily without type information (when the types are concrete), e.g. compare on master:
```julia
julia> show(Int8[1])
Int8[1]

julia> show(Dict(Int8(1)=>1))
Dict(1=>1)
```
I.e. compared to arrays, we loose information when showing a `Dict`. I don't find this ideal, but this is not directly related to this PR.

**Pull request merged**